### ISSUE TITLE
spaceless feature should be a tag rather than a filter

### ISSUE THREAD

**Issue by jsha** [Keats/tera#675]
HIya! I filed #641 earlier this year, and @Urgau quickly sent a pull request #643 to address it. Thanks so much! I should have been paying closer attention at the time, but the implementation differs from Django's in an important way. Django's spaceless (https://docs.djangoproject.com/en/3.2/ref/templates/builtins/#spaceless) is a tag, not a template. It's an operation you want to apply to a template at parsing time, rather than a value at interpolation time.

I know this feature has been released already, but would consider changing this to a tag? As @Keats mentions in #641, doing this as a filter doesn't really work, because there can be `>` and `<` inside quoted strings inside HTML - you'd have to be HTML-aware. However, doing this as a tag at the template level works reasonably well, because the contents of the template are static and predictable.

**Comment by Urgau** [comment]
@jsha Couldn't you use it with the `filter` tag ?

```tera
{% filter spaceless %}
    <p>
        <a href="foo/">Foo</a>
    </p>
{% endfilter %}
```

**Comment by Urgau** [comment]
> I guess it would become `<ahref="foo/">Foo</a>`?

No, spaceless only remove spaces between `>` and `<`, so the example would become: `"\n    <p><a href=\"foo/\">Foo</a></p>\n\n"` (the `\n` comes from the p tag, ie no trimming).

**Comment by winstonewert** [comment]
Is there a reason to do this over running a minifier on the end result html?

**Comment by jsha** [comment]
Just that it's nicer to emit directly the output we want rather than spending CPU to re-parse and re-emit. However, the project I wanted this for (rustdoc) has since moved to Askama for templating, so I'm happy to close this out. Though it's worth considering if the existing spaceless feature should be deprecated since it's can produce a subtle bug in cases like this:

```
{% filter spaceless %}
<a href="https://example.com" title="<x>      <y>">example</a> 
{% endfilter %}
```

Which I believe would be emitted as:

```
<a href="https://example.com" title="<x><y>">example</a> 
```

**Comment by winstonewert** [comment]
I will note that the implementation of this renders the html and then runs the filter on it, so you don't really save the re-parsing/re-emitting CPU. It would certainly be better to run a proper html minifier.

So, I'll suggest this really ought to be deprecated because its a bad idea all around.

**Comment by Keats** [comment]
> So, I'll suggest this really ought to be deprecated because its a bad idea all around.

Likely to be removed for v2.

**Comment by Keats** [comment]
Closing as part of the Tera v2 release. If this is still relevant in v2, please reopen — see the 
  migration guide: https://github.com/Keats/tera/blob/master/MIGRATION.md
